pkg/trace/obfuscate/json.go

```diff
@@ -58,9 +79,21 @@ func (p *jsonObfuscator) setKey() {
 	p.wiped = false
 }
 
+func trimSideQuotes(quotedString []byte) []byte {
+	if len(quotedString) > 0 && quotedString[0] == '"' {
+		quotedString = quotedString[1:]
+	}
+	if len(quotedString) > 0 && quotedString[len(quotedString)-1] == '"' {
+		quotedString = quotedString[0 : len(quotedString)-1]
+	}
+	return quotedString
+}
+
 func (p *jsonObfuscator) obfuscate(data []byte) (string, error) {
 	var out strings.Builder
-	buf := make([]byte, 0, 10) // recording key token
+
+	keyBuf := make([]byte, 0, 10) // recording key token
```

Comment:
We're going to need some numbers with `benchcmp` to see how this compares to the previous implementation. We know from profiles that on heavy sql spans this does weigh down on performance.

Reply:
Here it is: 

```
benchcmp old.txt new.txt
benchcmp is deprecated in favor of benchstat: https://pkg.go.dev/golang.org/x/perf/cmd/benchstat
ignoring BenchmarkObfuscateJSON/5003-8: before has 1 instances, after has 0
ignoring BenchmarkObfuscateJSON/437-8: before has 1 instances, after has 0
ignoring BenchmarkObfuscateJSON/671-8: before has 1 instances, after has 0
```

And the benchmarks themselves:
```
cat old.txt
goos: darwin
goarch: amd64
pkg: github.com/DataDog/datadog-agent/pkg/trace/obfuscate
BenchmarkObfuscateJSON/5003-8         	   24805	     47720 ns/op	   22120 B/op	     203 allocs/op
BenchmarkObfuscateJSON/437-8          	  191492	      6396 ns/op	    2952 B/op	      48 allocs/op
BenchmarkObfuscateJSON/671-8          	  175016	      6952 ns/op	    3016 B/op	      42 allocs/op
PASS
ok  	github.com/DataDog/datadog-agent/pkg/trace/obfuscate	4.387s

cat new.txt
goos: darwin
goarch: amd64
pkg: github.com/DataDog/datadog-agent/pkg/trace/obfuscate
BenchmarkObfuscateJSON/806-8         	   54447	     22840 ns/op	   10174 B/op	      80 allocs/op
BenchmarkObfuscateJSON/566-8         	   69775	     17706 ns/op	    8337 B/op	      67 allocs/op
BenchmarkObfuscateJSON/31-8          	  125083	     10101 ns/op	    5865 B/op	      35 allocs/op
PASS
ok  	github.com/DataDog/datadog-agent/pkg/trace/obfuscate	4.428s
```

This shouldn't have any impact on regular APM json obfuscation because the new value transformer code I added will only run when a value transformer is configured, which is only happening in the sql obfuscation code which is called from python checks. 

It can be configured to be used for APM though, so if we do want to obfuscate sql within json then we'll be able to do so.

Reply:
Well it looks like everything is worse.

Reply:
Looks like it's all in the initialization. If we move the obfuscator initialization outside of the benchmark then the benchmarks aren't so different:

```
benchcmp is deprecated in favor of benchstat: https://pkg.go.dev/golang.org/x/perf/cmd/benchstat
benchmark                                          old ns/op     new ns/op     delta
BenchmarkObfuscateJSON/elasticsearch.body-8        48500         53455         +10.22%
BenchmarkObfuscateJSON/elasticsearch.body#01-8     6049          6368          +5.27%
BenchmarkObfuscateJSON/elasticsearch.body#02-8     6514          7253          +11.34%
BenchmarkObfuscateJSON/elasticsearch.body#03-8     6486          7418          +14.37%
BenchmarkObfuscateJSON/elasticsearch.body#04-8     1844          2066          +12.04%
BenchmarkObfuscateJSON/elasticsearch.body#05-8     1804          2121          +17.57%
BenchmarkObfuscateJSON/elasticsearch.body#06-8     1807          2112          +16.88%
BenchmarkObfuscateJSON/elasticsearch.body#07-8     1831          2052          +12.07%
BenchmarkObfuscateJSON/elasticsearch.body#08-8     748           802           +7.22%
BenchmarkObfuscateJSON/elasticsearch.body#09-8     918           1268          +38.13%
BenchmarkObfuscateJSON/elasticsearch.body#10-8     975           1084          +11.18%
BenchmarkObfuscateJSON/elasticsearch.body#11-8     1514          1896          +25.23%
BenchmarkObfuscateJSON/elasticsearch.body#12-8     1099          1250          +13.74%
BenchmarkObfuscateJSON/elasticsearch.body#13-8     1491          1685          +13.01%

benchmark                                          old allocs     new allocs     delta
BenchmarkObfuscateJSON/elasticsearch.body-8        195            195            +0.00%
BenchmarkObfuscateJSON/elasticsearch.body#01-8     39             39             +0.00%
BenchmarkObfuscateJSON/elasticsearch.body#02-8     33             33             +0.00%
BenchmarkObfuscateJSON/elasticsearch.body#03-8     33             33             +0.00%
BenchmarkObfuscateJSON/elasticsearch.body#04-8     16             16             +0.00%
BenchmarkObfuscateJSON/elasticsearch.body#05-8     16             16             +0.00%
BenchmarkObfuscateJSON/elasticsearch.body#06-8     16             16             +0.00%
BenchmarkObfuscateJSON/elasticsearch.body#07-8     16             16             +0.00%
BenchmarkObfuscateJSON/elasticsearch.body#08-8     9              9              +0.00%
BenchmarkObfuscateJSON/elasticsearch.body#09-8     9              9              +0.00%
BenchmarkObfuscateJSON/elasticsearch.body#10-8     9              9              +0.00%
BenchmarkObfuscateJSON/elasticsearch.body#11-8     14             14             +0.00%
BenchmarkObfuscateJSON/elasticsearch.body#12-8     11             11             +0.00%
BenchmarkObfuscateJSON/elasticsearch.body#13-8     14             14             +0.00%

benchmark                                          old bytes     new bytes     delta
BenchmarkObfuscateJSON/elasticsearch.body-8        21702         21698         -0.02%
BenchmarkObfuscateJSON/elasticsearch.body#01-8     2456          2456          +0.00%
BenchmarkObfuscateJSON/elasticsearch.body#02-8     2520          2520          +0.00%
BenchmarkObfuscateJSON/elasticsearch.body#03-8     2520          2520          +0.00%
BenchmarkObfuscateJSON/elasticsearch.body#04-8     712           712           +0.00%
BenchmarkObfuscateJSON/elasticsearch.body#05-8     712           712           +0.00%
BenchmarkObfuscateJSON/elasticsearch.body#06-8     712           712           +0.00%
BenchmarkObfuscateJSON/elasticsearch.body#07-8     712           712           +0.00%
BenchmarkObfuscateJSON/elasticsearch.body#08-8     296           296           +0.00%
BenchmarkObfuscateJSON/elasticsearch.body#09-8     328           328           +0.00%
BenchmarkObfuscateJSON/elasticsearch.body#10-8     328           328           +0.00%
BenchmarkObfuscateJSON/elasticsearch.body#11-8     616           616           +0.00%
BenchmarkObfuscateJSON/elasticsearch.body#12-8     504           488           -3.17%
BenchmarkObfuscateJSON/elasticsearch.body#13-8     616           616           +0.00%
```

Reply:
Fixed here https://github.com/DataDog/datadog-agent/pull/5692/commits/0d73bfc16cd4184963480ca84c7e763d5dfc3e5b

Reply:
Cool, I think it looks good. Thanks